File: app/src/main/java/org/kiwix/kiwixmobile/bookmark/BookmarksPresenter.java
/*
 * Kiwix Android
 * Copyright (c) 2019 Kiwix <android.kiwix.org>
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 *
 */

package org.kiwix.kiwixmobile.bookmark;

import android.util.Log;
import io.reactivex.CompletableObserver;
import io.reactivex.Observable;
import io.reactivex.Scheduler;
import io.reactivex.SingleObserver;
import io.reactivex.disposables.Disposable;
import java.util.List;
import javax.inject.Inject;
import org.kiwix.kiwixmobile.base.BasePresenter;
import org.kiwix.kiwixmobile.data.DataSource;
import org.kiwix.kiwixmobile.di.PerActivity;
import org.kiwix.kiwixmobile.di.qualifiers.Computation;
import org.kiwix.kiwixmobile.di.qualifiers.MainThread;

@PerActivity
class BookmarksPresenter extends BasePresenter<BookmarksContract.View>
  implements BookmarksContract.Presenter {

  private final DataSource dataSource;
  private final Scheduler mainThread;
  private final Scheduler computation;
  private Disposable disposable;

  @Inject BookmarksPresenter(DataSource dataSource, @MainThread Scheduler mainThread,
    @Computation Scheduler computation) {
    this.dataSource = dataSource;
    this.mainThread = mainThread;
    this.computation = computation;
  }

  @Override
  public void loadBookmarks(boolean showFromCurrentBookmarks) {
    dataSource.getBookmarks(showFromCurrentBookmarks)
      .subscribe(new SingleObserver<List<BookmarkItem>>() {
        @Override
        public void onSubscribe(Disposable d) {
          if (disposable != null && !disposable.isDisposed()) {
            disposable.dispose();
          }
          disposable = d;
          compositeDisposable.add(d);
        }

        @Override
        public void onSuccess(List<BookmarkItem> bookmarks) {
          view.updateBookmarksList(bookmarks);
        }

        @Override
        public void onError(Throwable e) {
          Log.e("BookmarksPresenter", e.toString());
        }
      });
  }

  @Override
  public void filterBookmarks(List<BookmarkItem> bookmarks, String newText) {
    Observable.fromIterable(bookmarks)
      .filter(
        bookmark -> bookmark.getBookmarkTitle().toLowerCase().contains(newText.toLowerCase()))
      .toList()
      .subscribeOn(computation)
      .observeOn(mainThread)
      .subscribe(new SingleObserver<List<BookmarkItem>>() {
        @Override
        public void onSubscribe(Disposable d) {
          compositeDisposable.add(d);
        }

        @Override
        public void onSuccess(List<BookmarkItem> bookmarkList) {
          view.notifyBookmarksListFiltered(bookmarkList);
        }

        @Override
        public void onError(Throwable e) {
          Log.e("BookmarksPresenter", e.toString());
        }
      });
  }

  @Override
  public void deleteBookmarks(List<BookmarkItem> deleteList) {
    dataSource.deleteBookmarks(deleteList)
      .subscribe(new CompletableObserver() {
        @Override
        public void onSubscribe(Disposable d) {

        }

        @Override
        public void onComplete() {

        }

        @Override
        public void onError(Throwable e) {
          Log.e("BookmarksPresenter", e.toString());
        }
      });
  }
}

File: core/src/main/java/org/kiwix/kiwixmobile/core/bookmark/BookmarksPresenter.kt
/*
 * Kiwix Android
 * Copyright (c) 2019 Kiwix <android.kiwix.org>
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 *
 */
package org.kiwix.kiwixmobile.core.bookmark

import android.util.Log
import io.reactivex.Observable
import io.reactivex.Scheduler
import io.reactivex.disposables.Disposable
import org.kiwix.kiwixmobile.core.base.BasePresenter
import org.kiwix.kiwixmobile.core.data.DataSource
import org.kiwix.kiwixmobile.core.di.qualifiers.Computation
import org.kiwix.kiwixmobile.core.di.qualifiers.MainThread
import javax.inject.Inject

class BookmarksPresenter @Inject constructor(
  private val dataSource: DataSource,
  @param:MainThread private val mainThread: Scheduler,
  @param:Computation private val computation: Scheduler
) : BasePresenter<BookmarksContract.View>(),
  BookmarksContract.Presenter {
  private var disposable: Disposable? = null
  override fun loadBookmarks(showBookmarksCurrentBook: Boolean) {
    disposable?.takeIf { !it.isDisposed }?.dispose()
    dataSource.getBookmarks(showBookmarksCurrentBook)
      .subscribe(
        { view?.updateBookmarksList(it) },
        { Log.e("BookmarkPresenter", "Failed to load bookmarks", it) }
      ).let {
        compositeDisposable.add(it)
        disposable = it
      }
  }

  override fun filterBookmarks(bookmarksList: List<BookmarkItem>, newText: String) {
    compositeDisposable.add(Observable.fromCallable {
      bookmarksList.filter {
        it.bookmarkTitle.contains(newText, true)
      }
    }
      .subscribeOn(computation)
      .observeOn(mainThread)
      .subscribe(
        { view?.notifyBookmarksListFiltered(it) },
        { Log.e("BookmarkPresenter", "Failed to filter bookmark.", it) }
      )
    )
  }

  override fun deleteBookmarks(deleteList: List<BookmarkItem>) {
    dataSource.deleteBookmarks(deleteList)
      .subscribe({}, { Log.e("BookmarkPresenter", "Failed to delete bookmark", it) }
      )
  }
}
